test/extended/dr/machine_recover.go

```diff
@@ -0,0 +1,330 @@
+				// retry until the machine gets created
+				err = wait.PollImmediate(5*time.Second, 10*time.Minute, func() (bool, error) {
+					_, err := ms.Create(newMaster, metav1.CreateOptions{})
+					if errors.IsAlreadyExists(err) {
+						framework.Logf("Waiting for old machine object %s to be deleted so we can create a new one", replacedMaster.Name)
+						return false, nil
+					}
+					if err != nil {
+						return false, err
+					}
+					return true, nil
+				})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				framework.Logf("Wait for masters to join as nodes and go ready")
+				err = wait.Poll(30*time.Second, 30*time.Minute, func() (done bool, err error) {
+					defer func() {
+						if r := recover(); r != nil {
+							fmt.Println("Recovered from panic", r)
+						}
+					}()
+					nodes, err := oc.AdminKubeClient().CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/master="})
+					if err != nil {
+						return false, err
+					}
+					ready := countReady(nodes.Items)
+					if ready != len(masters) {
+						framework.Logf("%d master nodes still unready", len(masters)-ready)
+						return false, nil
+					}
+					return true, nil
+				})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				framework.Logf("Verify recreated master has healthy etcd endpoint")
+				// unhealthyEtcds = getUnhealthyEtcds(certsDir, masters)
+				// o.Expect(len(unhealthyEtcds)).To(o.BeNumerically("==", 0))
+			})
+	},
+	)
+})
+
+func deleteMachineCheckForRole(role string) {
+	config, err := framework.LoadConfig()
+	o.Expect(err).NotTo(o.HaveOccurred())
+	dynamicClient := dynamic.NewForConfigOrDie(config)
+	mhc := dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    "machine.openshift.io",
+		Version:  "v1beta1",
+		Resource: "machinehealthchecks",
+	}).Namespace("openshift-machine-api")
+	err = mhc.Delete("e2e-health-check-"+role, &metav1.DeleteOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+}
+
+func createMachineHealthCheckForRole(role string) {
+	config, err := framework.LoadConfig()
+	o.Expect(err).NotTo(o.HaveOccurred())
+	dynamicClient := dynamic.NewForConfigOrDie(config)
+	mhc := dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    "machine.openshift.io",
+		Version:  "v1beta1",
+		Resource: "machinehealthchecks",
+	}).Namespace("openshift-machine-api")
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "machine.openshift.io",
+		Version: "v1beta1",
+		Kind:    "machinehealthchecks",
+	})
+	err = unstructured.SetNestedField(u.Object, role, "spec", "selector", "machine.openshift.io/cluster-api-machine-role")
+	o.Expect(err).ToNot(o.HaveOccurred())
+	err = unstructured.SetNestedField(u.Object, role, "metadata", "name", "e2e-health-check-"+role)
+	o.Expect(err).ToNot(o.HaveOccurred())
+	err = unstructured.SetNestedField(u.Object, []interface{}{
+		map[string]string{
+			"type":    "Ready",
+			"timeout": "5m",
```

Comment:
Hrm, I don’t think the test is doing what we expect.  It takes 2m to run but the timeout is longer.  Can you run the test directly against a cluster and verify the expected flows occur?

Reply:
@smarterclayton I think we never waited for the worker node to get created after 5 min timeout. And we create the master machine in the test itself which was the reason for smaller time window. I added the code[1] to wait for worker node, hopefully, that would give test times closer to, if not more than, 5mins

1. https://github.com/openshift/origin/pull/24651/files#diff-a7ee9c3c62d07e787a70359b84c68fe7R184

Reply:
> 

[sig-cluster-lifecycle][Feature:DisasterRecovery][Disruptive] [Feature:NodeRecovery] Cluster should survive master and worker failure and recover with machine health checks [Serial] [Suite:openshift] | 5m19s
-- | --

https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/24651/pull-ci-openshift-origin-master-e2e-aws-disruptive/55

@smarterclayton PTAL, I can still run locally if need be, thanks.

Reply:
I think that’s what I expect.  Before we merge I want to merge a fix that ensures logs are preserved when we call openshift tests twice in a row (caught that debugging).

Reply:
sounds good, thanks